zerver/data_import/slack.py

```diff
@@ -463,21 +471,54 @@ def process_mpims(mpims: List[Dict[str, Any]]) -> None:
         mpims = []
     process_mpims(mpims)
 
-    for user in zerver_userprofile:
+    def get_user_recipient_id(username: str) -> int:
+        nonlocal recipient_id_count
+
+        if username not in added_recipient:
+            recipient_id = recipient_id_count
+            recipient = build_recipient(added_users[username], recipient_id, Recipient.PERSONAL)
+            added_recipient[username] = recipient_id
+            recipient_id_count += 1
+            zerver_recipient.append(recipient)
+        else:
+            recipient_id = added_recipient[username]
+        return recipient_id
+
+    def process_dms(dms: List[Dict[str, Any]]) -> None:
+        nonlocal recipient_id_count
+        nonlocal subscription_id_count
+
+        for dm in dms:
+            user_a = dm["members"][0]
+            user_b = dm["members"][1]
+
+            user_a_recipient = get_user_recipient_id(user_a)
+            user_b_recipient = get_user_recipient_id(user_b)
+
+            dm_members[dm["id"]] = (user_a, user_b)
+            sub1 = build_subscription(user_a_recipient, added_users[user_b], subscription_id_count)
+            sub2 = build_subscription(user_b_recipient, added_users[user_a], subscription_id_count + 1)
+
+            zerver_subscription.extend([sub1, sub2])
+            subscription_id_count += 2
+
+    try:
+        dms = get_data_file(slack_data_dir + '/dms.json')
+    except FileNotFoundError:
+        dms = []
+    process_dms(dms)
+
+    for username in added_users:
         # this maps the recipients and subscriptions
         # related to private messages
-        recipient = build_recipient(user['id'], recipient_id_count, Recipient.PERSONAL)
-        sub = build_subscription(recipient_id_count, user['id'], subscription_id_count)
-
-        zerver_recipient.append(recipient)
+        recipient_id = get_user_recipient_id(username)
+        sub = build_subscription(recipient_id, added_users[username], subscription_id_count)
```

Comment:
I'm a little confused by this restructuring.  It seems like we should do the construction of all the `zerver_recipient` objects from all the users in the organization in a loop first (recording the data needed to look them up when processing the DMs), and then process the DMs, rather than having this `get_user_recipient_id` function that we call from 2 code paths (it's be more readable that way).